llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp

```diff
@@ -0,0 +1,574 @@
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitSymbol(const MachineInstr *MI,
+                                                     MCSymbol *Symbol) {
+  unsigned Opc = MI->getOpcode();
+  assert(Opc == EVM::DATASIZE ||
+         Opc == EVM::DATAOFFSET && "Unexpected symbol instruction");
+  StackHeight += 1;
+  // This is codegen-only instruction, that will be converted into PUSH4.
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::getStackOpcode(Opc)))
+                   .addSym(Symbol);
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitConstant(uint64_t Val) {
+  emitConstant(APInt(256, Val));
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitLabelReference(
+    const MachineInstr *Call) {
+  assert(Call->getOpcode() == EVM::FCALL && "Unexpected call instruction");
+  StackHeight += 1;
+  auto [It, Inserted] = CallReturnSyms.try_emplace(Call);
+  if (Inserted)
+    It->second = MF.getContext().createTempSymbol("FUNC_RET", true);
+  auto NewMI =
+      BuildMI(*CurMBB, CurMBB->end(), DebugLoc(), TII->get(EVM::PUSH_LABEL))
+          .addSym(It->second);
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitFuncCall(const MachineInstr *MI,
+                                                       const GlobalValue *Func,
+                                                       int StackAdj,
+                                                       bool WillReturn) {
+  assert(MI->getOpcode() == EVM::FCALL && "Unexpected call instruction");
+  assert(CurMBB == MI->getParent());
+
+  // PUSH_LABEL technically increases the stack height on 1, but we don't
+  // increase it explicitly here, as the label will be consumed by the following
+  // JUMP.
+  StackHeight += StackAdj;
+
+  // Create pseudo jump to the callee, that will be expanded into PUSH and JUMP
+  // instructions in the AsmPrinter.
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::PseudoCALL))
+                   .addGlobalAddress(Func);
+
+  // If this function returns, we need to create a label after JUMP instruction
+  // that is followed by JUMPDEST and this is taken care in the AsmPrinter.
+  // In case we use setPostInstrSymbol here, the label will be created
+  // after the JUMPDEST instruction, which is not what we want.
+  if (WillReturn)
+    NewMI.addSym(CallReturnSyms.at(MI));
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitRet(const MachineInstr *MI) {
+  assert(MI->getOpcode() == EVM::RET && "Unexpected ret instruction");
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::PseudoRET));
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitUncondJump(
+    const MachineInstr *MI, MachineBasicBlock *Target) {
+  assert(MI->getOpcode() == EVM::JUMP &&
+         "Unexpected unconditional jump instruction");
+  assert(StackHeight >= 0);
```

Comment:
Redundant as we check `StackHeight` every time we decrease it.

Reply:
Removed this assert.